### PR TITLE
[Feature/extensions] UserIdentity object for OpenSearch permissions and access

### PR DESCRIPTION
**Pull request by peternied**
### Description
Adding basic identity objects for OpenSearch, unifying all identity details within core.
 
### Issues Resolved
* Related https://github.com/opensearch-project/OpenSearch/issues/3846
* Related https://github.com/opensearch-project/opensearch-sdk/issues/37
* Design documentation https://github.com/opensearch-project/opensearch-sdk/blob/main/SECURITY.md#projects
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@peternied curious, what are the benefits of introducing own identity hierarchy vs relying on JDK's `java.security.Principal` [1]?

[1] https://docs.oracle.com/javase/8/docs/api/java/security/Principal.html

**Comment by peternied** [comment]
> @peternied curious, what are the benefits of introducing own identity hierarchy vs relying on JDK's `java.security.Principal` [1]?
> 
> [1] https://docs.oracle.com/javase/8/docs/api/java/security/Principal.html

Principal is tightly coupled with the SecurityManager implemenation, the method 'boolean implies(Subject)` [2] - which is not an interface, but a final class.  There is flexibility in `Subject`, but there is many implied behaviors that might not be correct for OpenSearch's identity systems, such more patterns that depend on SecurityManager via AccessController.doPrivileged(...) calls [3].

Around this space there are several implementations of Principal that are depreciated that would temp usage in the codebase such as `Identity` /  `IdentityScope`.  Of the non-depreciated implementations most are related to certification management that does not seem to align well for user centric identity scenarios.

I would rather start with less flexibility to ensure fewer considerations around security practices.

[2] https://docs.oracle.com/javase/8/docs/api/java/security/Principal.html#implies-javax.security.auth.Subject-
[3] https://docs.oracle.com/javase/8/docs/api/javax/security/auth/Subject.html#doAsPrivileged-javax.security.auth.Subject-java.security.PrivilegedAction-java.security.AccessControlContext-

**Comment by reta** [comment]
> Principal is tightly coupled with the SecurityManager implemenation, the method `boolean implies(Subject)` [2] - which is not an interface, but a final class. There is flexibility in `Subject`, but there is many implied behaviors that might not be correct for OpenSearch's identity systems, such more patterns that depend on SecurityManager via AccessController.doPrivileged(...) calls [3].

I believe the dependency on `Subject` and `SecurityManager` is not really a concern since the whole OpenSearch core codebase depends on that (also, the `implies` is the default method, so we could leave it as is). AFAIK, most (if not all) security frameworks in JVM ecosystem [1], [2], [3] are able to deal with  `java.security.Principal` out of the box, another benefit to rely on standard library.

> I would rather start with less flexibility to ensure fewer considerations around security practices, what do you think?

Using existing `java.security.Principal` looks like the most straightforward option to get started, at least less abstractions to invent. I does not require to implement more than we need (name / equals / hashcode). Just my 5 cents.

[1] https://github.com/spring-projects/spring-security
[2] https://github.com/apache/shiro
[3] https://github.com/apache/cxf

File: server/src/main/java/org/opensearch/identity/Identity.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+package org.opensearch.identity;
+
+/**
+ * The distinguishing character or personality of an entity within OpenSearch.
+ */
+public interface Identity {
+
+    /** A durable, reusable identifier of this identity */
+ 
+    String getId();
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/Identity.java:21]: If we're creating our own Identity object, we should have fields that we can deserialize with Jackson to follow at least the minimal user representation per [SCIM](https://www.simplecloud.info/).  From [RFC 7643](https://datatracker.ietf.org/doc/html/rfc7643#section-8.1):
```json
{
  "schemas": ["urn:ietf:params:scim:schemas:core:2.0:User"],
  "id": "2819c223-7f76-453a-919d-413861904646",
  "userName": "bjensen@example.com",
  "meta": {
    "resourceType": "User",
    "created": "2010-01-23T04:56:22Z",
    "lastModified": "2011-05-13T04:42:34Z",
    "version": "W\/\"3694e05e9dff590\"",
    "location":
     "https://example.com/v2/Users/2819c223-7f76-453a-919d-413861904646"
  }
}

            Figure 3: Example Minimal User JSON Representation
```
So I'd suggest at least getters for `Set<String>` schemas, `String` id, `String` userName, and `Map<String,String>` meta.

But rather than code it ourselves, we could import a dependency that implements the standard.  See http://www.simplecloud.info/#Implementations2 and scan the Open Source (AL2).  I've seen [WSO2 Charon](https://github.com/wso2/charon) used elsewhere.  [Apache Syncope](https://syncope.apache.org/) is ASF-maintained.

They might be overkill, but would provide room to grow.  I'd recommend at least taking a look at how easy it is to generate a simple identity using one of those dependencies.

File: server/src/main/java/org/opensearch/identity/UserIdentity.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * An authenticated user in the system.
+ */
+public class UserIdentity implements Identity {
+
+    /** No identity information or identity information is unavaliable */
+    public static final AnonymousUser ANONYMOUS = new AnonymousUser();
+
+    /** A persistant identifier that is unqiue to the user */
+    private String id;
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/identity/UserIdentity.java:25]: Better name here 
```suggestion
    private String UserId;
```

File: server/src/main/java/org/opensearch/identity/package-info.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/** OpenSearch Identity Classes. */
+package org.opensearch.identity;
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/package-info.java:10]: I like this package name.

File: server/src/main/java/org/opensearch/identity/UserIdentity.java (hunk 2 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/identity/UserIdentity.java:25]: Also, are we planning to have id as a String?

File: server/src/main/java/org/opensearch/identity/UserIdentity.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.List;
+
+/**
+ * An authenticated user in the system.
+ */
+public class UserIdentity implements Identity {
+
+    /** No identity information or identity information is unavaliable */
+    public static final AnonymousUser ANONYMOUS = new AnonymousUser();
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/UserIdentity.java:22]: I like having a constant for a "this user is completely untrusted" for many reasons.

File: server/src/main/java/org/opensearch/identity/UserIdentity.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    /** A durable and reusable identifier of this user */
+    public String getId() {
+        return this.id;
+    }
+
+    /** No identity information or identity information is unavaliable */
+    public static final class AnonymousUser extends UserIdentity {
+        static final String ID = "Anonymous";
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/UserIdentity.java:38]: I think an empty string may be better to use for Anonymous.  Also in case we want to use a format like UUID for this, we'd not want to create a constant that conflicts with it. 

Trivia: did you know there's a hex color code for Anonymous?[ `#817A6E`](https://encycolorpedia.com/817a6e).   `UUID.fromBytes()` with repetitions of those byte values would be a fun constant.

File: server/src/test/java/org/opensearch/identity/UserIdentityTests.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import org.opensearch.identity.UserIdentity.AnonymousUser;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.hamcrest.core.IsNot.not;
+
+public class UserIdentityTests {
+
+    public void testGetIds() {
+        final String idOfSally = "Sally";
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/identity/UserIdentityTests.java:23]: Alice might be annoyed that Bob is now hanging out with Sally.

More seriously, I'd like to get away from using names as "ID" strings.  Particularly if we use SCIM there's a separate name field/fields.

File: server/src/test/java/org/opensearch/identity/UserIdentityTests.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            public String getId() { return null; }
+        };
+    
+        final Identity otherIdentity = new Identity() {
+            @Override
+            public String getId() { return "other"; }
+        };
+    
+        assertThat(UserIdentity.ANONYMOUS, equalTo(UserIdentity.ANONYMOUS));
+        assertThat(UserIdentity.ANONYMOUS, not(equalTo(blankIdentity)));
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/identity/UserIdentityTests.java:59]: I'm curious why we should allow null, but not make null the anonymous identity.

File: server/src/main/java/org/opensearch/identity/UserIdentity.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        protected AnonymousUser() {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/identity/UserIdentity.java:39]: The `protected` modifier doesn't match `final` on the class. This class must not be allowed to be inherited.

**Comment by peternied** [comment]
Thanks for the great feedback @dbwiddis @owaiskazi19 @reta - I am doing a deeper dive onto [Apache Shiro](https://github.com/apache/shiro) that I hadn't come across before, I'll circle back around to the outstanding feedback...

FYI @opensearch-project/security

File: server/src/main/java/org/opensearch/identity/UserIdentity.java (hunk 2 above)
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/identity/UserIdentity.java:25]: I prefer `id` to match SCIM field definition.

**Comment by dbwiddis** [comment]
> It does not require to implement more than we need (name / equals / hashcode). Just my 5 cents.

@reta @peternied from the standpoint of a review of this particular PR I would agree that `Principal` (and its subinterface `Group`) provide what we need to implement _internally to OpenSearch_, which is a unique identity of a user (or a group of which the caller is a member).

I highlighted the internal representation as separate from an external representation. From the standpoint of extensions, user identity information will be shared via REST APIs where we will be implementing a larger scoped Identity and Access Management (IAM) solution involving OAuth2, OpenID Connect, and SCIM.

If we want to take a more complete identity from SCIM and dumb it down to a single "name" internally (so long as that name is unique!) to gain some benefit, that's reasonable, particularly if it's already an implemented, working solution.  However if we're designing from scratch, deserializing JSON into an object matching SCIM would be much simpler and convey a richer identity scope than simply a name and class.

**Comment by dbwiddis** [comment]
> I am doing a deeper dive onto [Apache Shiro](https://github.com/apache/shiro) that I hadn't come across before

Spent a bit of time looking at that myself tonight.  Looks useful from a user/roles/etc. back-end standpoint.

I'll repeat some comments that I've made on other issues that I don't recall where they are but want to bring some context to this review:
 - The REST API is somewhat the "boundary" where we may have a different representation externally than we do internally
 - Internally we need details on users/roles/etc.   All of this can be tied to a single unique "ID".
 - Externally we will be using an access token associated with that ID.   Presenting the access token (only) at the REST API boundary will unlock all the internal details.
 - The point where I'm interested in this PR and user representation is at the point we provide this access token.  We'll want to provide access to an authenticated "identity"
   - We can do the authentication ourselves (user/password, certificate-based) or via SSO or federated access. (Eventual future plan. We'll start slow.)
   - Using standardized identity representations (SCIM) and methods (Open ID Connect) is key to making sure we enable this capability, even if it's a bit more than we currently need.  I think this is the right place to add at least the basics of this.

**Comment by reta** [comment]
@peternied @dbwiddis it strikes me (from the comments above) we need to spend some efforts on design first in order to understand what we would like to achieve and how to get there. The only design piece I found is [1] which is just a high level chart without any details sketched out before coding. WDYT?

[1] https://github.com/opensearch-project/OpenSearch/issues/3846#issuecomment-1185994410

**Comment by dbwiddis** [comment]
> @peternied @dbwiddis it strikes me (from the comments above) we need to spend some efforts on design first in order to understand what we would like to achieve and how to get there. 

I agree.  On the one hand, I don't think the User Identity is that difficult to set up an initial representation.  However, we do need a much more detailed design on how we map identity to access and roles (and possibly tenants).

**Comment by saratvemulapalli** [comment]
Lots good feedback (I spent a bunch of time reading thanks @reta @dbwiddis ), @peternied as we spend sometime researching. May be move this to a draft state?

**Comment by peternied** [comment]
Dood idea @saratvemulapalli done, I am really warming up to using Java security principal, but I'll have that assembled design when I can push some pixels together with a write up

File: server/src/main/java/org/opensearch/identity/Identity.java (hunk 1 above)
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/Identity.java:21]: Good notion following the standard format, while I didn't include serialization in this change I am onboard that our solution should follow suite.

Your instincts to leverage an external library are on point - this is the path I'll redirect to

File: server/src/main/java/org/opensearch/identity/UserIdentity.java (hunk 5 above)
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/UserIdentity.java:38]: I like your UUID suggestion, great easter egg, as it would limit collisions by accident with a pseudo-reserved identity type

File: server/src/test/java/org/opensearch/identity/UserIdentityTests.java (hunk 6 above)
Comment (peternied) [reply server/src/test/java/org/opensearch/identity/UserIdentityTests.java:23]: I think that tests are easier to read with these simpler strings even if the standard representation is a uuid.  Note; another great reason to use a library not having to author these tests.

File: server/src/test/java/org/opensearch/identity/UserIdentityTests.java (hunk 7 above)
Comment (peternied) [reply server/src/test/java/org/opensearch/identity/UserIdentityTests.java:59]: The notion was to avoid the risky pattern `if (obj == null) { do something }`, but when considering the where/how this code is written is a great example to say plugins shouldn't be doing this directly - they shouldn't access the user object at all in a security centric concept.  If we were going to move forward with this implementation I would step this back.

**Comment by peternied** [comment]
Thanks for waiting while I good distracted in other issues, I'm going to be moving forward with a prototype of Shiro and circle back with findings.  For the moment no new `org.opensearch.identity` package, but maybe soon.
